docs: basic overview of git workflow

exu3 opened the pull request:
We created a markdown file that briefly covers the following: clone, pull, commit, push, pull request, and merge conflicts. 

Fixes #6

julialding committed:
added 3 tutorials & git dir

julialding committed:
fixed formatting

julialding committed:
Merge branch 'main' of https://github.com/ThinkFiveable/open into git-tutorials

julialding committed:
fixed minor errors

exu3 committed:
docs:combine git tutorials into one file

Combined git tutorials into one markdown file and created a TOC

exu3 committed:
fix:remove words that don't belong

exu3 committed:
docs: add pictures to visually show steps

exu3 committed:
docs: combined steps into one document

exu3 committed:
docs: combined steps into one document

exu3 committed:
docs: combined steps into one document

awlevin commented:
Oops didn’t mean to mark this as ready for review! Sorry bout that, the button was just so big and right where I was scrolling on mobile 😅

awlevin commented:
This looks fantastic so far!! Great work 😄

sarthaktexas committed:
Merge branch 'main' into git-tutorials-2

zaida04 commented on git-tutorials/git.md:
```diff
@@ -0,0 +1,55 @@
+`git clone <https address here>`
+
+For example:
+`git clone https://github.com/ThinkFiveable/open fiveable-open`
```
```suggestion
`git clone https://github.com/YOURUSERNAMEHERE/open.git fiveable-open`
```

Edit: I originally had this edit change it to: `git clone https://github.com/ThinkFiveable/open.git fiveable-open`, but then I wrote https://github.com/ThinkFiveable/open/pull/31#issuecomment-808788686

zaida04 commented on git-tutorials/git.md:
```diff
@@ -0,0 +1,55 @@
+You will need to use `git pull` to fetch the latest changes from another repository or local branch. When you run `git pull` in your terminal, remote content will be downloaded and immediately change the local state of your repository to match the new content. It is strongly recommended to not `git pull` when you have uncommitted changes. 
+
+In the event that you do have uncommitted changes and want to `git pull`, you will need to do the following in your command line.
+First, `git stash` which "stashes" your changes away temporarily. Then, `git pull` to fetch the latest changes. Next, `git stash pop` to apply the changes back onto your local copy.
```
```suggestion
First, `git stash` which "stashes" your changes away temporarily. Then, `git pull` to fetch the latest changes. Next, `git stash apply` to apply the changes back onto your local copy, and then `git stash drop` to remove those changes from the stash stack.
```
https://codingkilledthecat.wordpress.com/2012/04/27/git-stash-pop-considered-harmful/.

In short, unexpected behavior can occur in the event of pull/merge conflicts, which means git stash pop won't actually remove from the stack like expected. IMO it'd be better to be consistent to begin with than to have to teach edge cases like that.

zaida04 commented on git-tutorials/git.md:
```diff
@@ -0,0 +1,55 @@
+    + [Pull Request](#pull-request)
+
+# Getting Started with Git/GitHub
+To use Git, you'll need to download it [here](https://git-scm.com/downloads).
```
```suggestion
To use Git, you'll need to download it [here](https://git-scm.com/downloads).

Be sure to set your username and email before making contributions.
* `git config --global user.name "Your Name"`
* `git config --global user.email "youremail@yourdomain.com"`

zaida04 commented on git-tutorials/git.md:
```diff
@@ -0,0 +1,55 @@
+### Commit
+After making your changes, you will use terminal commands to commit them. Our commit message needs to explain what we've done. This message may be useful if we want to check our change history in the future, and it can also be useful to other contributors.
+
+The syntax for a commit message is as follows `git commit -m "your commit message here"`.
```
```suggestion
The syntax for a commit message is as follows `git commit -m "your commit message here"`.
For our repository, please ensure your commit messages follow [this standard](https://github.com/ThinkFiveable/open/blob/main/CONTRIBUTING.md#commiting)
```

zaida04 commented:
In addition to the changes above, there was one problem that I noticed will raise an issue later.

The clone URL that you provide will set the main remote url to the *main* repository. This will cause an issue when someone tries to do `git push`, as git will attempt to push to **this** repository, instead of a fork of their own repository. 

## So how do we solve this??
We need a `fork` section of this guide which teaches people how to fork the repository and then we need to edit the clone section appropriately to use THEIR forks URL. It will look something like `git clone https://github.com/URNAMEHERE/open`.

zaida04 commented:
And (last comment I swear 😛), we should encourage people to switch to a new branch when they make new changes instead of commiting to their master branch. 

## Why?
Say you commit to the master branch of your fork. You go ahead and make a PR, but oh no! Your PR was denied and was closed. You now have a problem, **your master branch now has a bunch of commits that don't exist on the upstream repo. Your fork is now out of sync with the upstream repo.** You *could* just revert those commits, but now you have a bunch of revert commits polluting your commit history, and those will be added to the upstream repo if you make another PR.

Sounds like a lot right? It's better to prevent all of this from happening in the beginning by just **making changes to a new branch every time**. So say I want to add a bunch of badges to the readme, i'd make a branch `chore/readme-branch` or `chore/patch-1`.

## So how would we solve this?
Add a `checkout` section to this guide that shows people how to switch to a new branch, create a new branch, and whatnot. 
*Also a nice little trick is `git switch branch-name` and `git switch -c branch-name`. Say you made a bunch of commits, but you forgot to switch branches. Instead of stashing and applying or discarding and redoing, you can just do the first or second command. The differences between the two are the first one will switch only to an existing branch, while the second one will create the branch if it doesn't exist. In the end, these both will **move** your already done commits to this new branch. Ref: https://git-scm.com/docs/git-switch

exu3 committed:
docs: add details, branching, and fork section

Added suggested changes and new sections including branching, forking, and setting user/email with `git config`

zaida04 commented on git-tutorials/git.md:
```diff
@@ -0,0 +1,78 @@
+`cd Documents`
+
+To clone our repository, we use the following command in our terminal:
+`git clone <https address here>`
```
```suggestion
`git clone <https address here> [cloned-directory-name]`
```

zaida04 commented on git-tutorials/git.md:
```diff
@@ -0,0 +1,78 @@
+### Pull
+You will need to use `git pull` to fetch the latest changes from another repository or local branch. When you run `git pull` in your terminal, remote content will be downloaded and immediately change the local state of your repository to match the new content. It is strongly recommended to not `git pull` when you have uncommitted changes. 
+
+In the event that you do have uncommitted changes and want to `git pull`, you will need to do the following in your command line. First, `git stash` which "stashes" your changes away temporarily. Then, `git pull` to fetch the latest changes. Next, `git stash apply` to apply the changes back onto your local copy, and then `git stash drop` to remove those changes from the stash stack. This method i ...
```
```suggestion
In the event that you do have uncommitted changes and want to `git pull`, you will need to do the following in your command line. First, `git stash` which "stashes" your changes away temporarily. Then, `git pull` to fetch the latest changes. Next, `git stash apply` to apply the changes back onto your local copy, and then `git stash drop` to remove those changes from the stash stack. 
```
💀 I think there was a mixup, `git stash pop` is what is not recommended and what'll cause unexpected behavior. `git stash apply` & `git stash drop` are the alternative.

zaida04 reviewed (changes requested):
Other than these two changes, this PR looks good to go! :D

exu3 committed:
docs: apply suggestions from code review

Co-authored-by: Zaid <nico.03727+github@gmail.com>

Commit:
docs: update branch to master section

Commit:
Merge branch 'main' into git-tutorials-2

exu3 commented on git-tutorials/git.md:
```diff
@@ -70,6 +70,17 @@ If you click that button, you will be asked to give your pull reqest a name and
+- Add our repository as a remote URL (`git remote add upstream https://github.com/ThinkFiveable/open.git`)
+- Make sure you are on the branch you want to merge master into (`git checkout PR_BRANCH_NAME`)
+- Make sure you have the latest upstream changes (`git fetch upstream`)
+- Merge the upstream master branch into your PR branch (`git merge upstream/master`). At this stage you might run into what are called **MERGE CONFLICTS. You must resolve these before pushing to your fork.** See: [Merge Conflicts](#merge-conflicts)
```
Should it be `git merge upstream/main`?

zaida04 replied:
Yeah totally forgot we use main

zaida04 committed:
Update git.md

sarthaktexas commented on git-tutorials/git.md:
```diff
@@ -0,0 +1,89 @@
+### Pull
+You will need to use `git pull` to fetch the latest changes from another repository or local branch. When you run `git pull` in your terminal, remote content will be downloaded and immediately change the local state of your repository to match the new content. It is strongly recommended to not `git pull` when you have uncommitted changes. 
+
+In the event that you do have uncommitted changes and want to `git pull`, you will need to do the following in your command line. First, `git stash` which "stashes" your changes away temporarily. Then, `git pull` to fetch the latest changes. Next, `git stash apply` to apply the changes back onto your local copy, and then `git stash drop` to remove those changes from the stash stack. 
```
why don't you just do git stash pop??

zaida04 replied:
https://github.com/ThinkFiveable/open/pull/31#discussion_r602780056
